vivisect/impemu/platarch/arm.py

```diff
@@ -1,14 +1,292 @@
+import sys
+import envi
+import logging
 import envi.archs.arm as e_arm
+
+import vivisect
 import vivisect.impemu.emulator as v_i_emulator
 
+import visgraph.pathcore as vg_path
+from envi.archs.arm.regs import *
+
+logger = logging.getLogger(__name__)
+verbose = True
+
 class ArmWorkspaceEmulator(v_i_emulator.WorkspaceEmulator, e_arm.ArmEmulator):
 
     taintregs = [ x for x in range(13) ]
 
     def __init__(self, vw, logwrite=False, logread=False):
         e_arm.ArmEmulator.__init__(self)
         v_i_emulator.WorkspaceEmulator.__init__(self, vw, logwrite=logwrite, logread=logread)
+        self.setMemArchitecture(envi.ARCH_ARMV7)
+
+    def setThumbMode(self, thumb=1):
+        self.opcache = {}
+        e_arm.ArmEmulator.setThumbMode(self, thumb)
+
+    def setArmMode(self, arm=1):
+        self.opcache = {}
+        e_arm.ArmEmulator.setArmMode(self, arm)
+
+    def parseOpcode(self, va, arch=envi.ARCH_DEFAULT):
+        '''
+        Caching version.  
+        
+        We can make an opcode *faster* with the workspace because of
+        getByteDef etc... use it.
+
+        Made for ARM, because envi.Emulator doesn't understand the Thumb flag
+        '''
+        op = self.opcache.get(va)
+
+        if op == None:
+            tmode = self.getFlag(PSR_T_bit)
+            if arch == envi.ARCH_DEFAULT:
+                arch = (envi.ARCH_ARMV7, envi.ARCH_THUMB)[tmode]
+
+            op = envi.archs.arm.emu.ArmEmulator.parseOpcode(self, va, arch=arch)
+            self.opcache[va] = op
+        return op
+
+    def stepi(self):
+        # NOTE: when we step, we *always* want to be stepping over calls
+        # (and possibly import emulate them)
+        starteip = self.getProgramCounter()
+
+        # parse out an opcode
+        tmode = self.getFlag(PSR_T_bit)
```

Comment:
normally i'd agree with you.  this is an oddity of ARM/THUMB.  i'm not sure how to generalize this with a hooking api without tanking performance or just being odd.  for the sake of performance, i reimplemented the function instead.  if you have suggestions that might make sense, i'd love to chat about them.

Reply:
code duplication is never fun.  agreed.  come up with an alternative that doesn't cost significant cycles during emulation, and perhaps we can rework how all of the WorkspaceEmulators work.  however, for the tight emulation/disassembly loops i tend to prefer to duplicate rather than incur additional calls.